Handle PlayStation platforms

objectx opened the pull request:
<!--
Please do not submit pull requests changing the `version.hpp`
or the single-include `catch.hpp` file, these are changed
only when a new release is made.

Before submitting a PR you should probably read the contributor documentation
at docs/contributing.md. It will tell you how to properly test your changes.
-->


## Description
This PR tries to resolve the following problems when compiling Catch2 for the Playstation platform.
 1. No POSIX signals
 2. No `getenv`
 3. No `gmtime_r`
     * `gmtime_s` exists but using [C11's definition](https://en.cppreference.com/w/c/chrono/gmtime) 😢

objectx committed:
feat: Handle PlayStation platforms

horenmar commented on src/catch2/internal/catch_test_case_registry_impl.cpp:
```diff
@@ -89,7 +89,7 @@ namespace Catch {
             return *lhs < *rhs;
         };
-        std::set<TestCaseInfo const*, decltype(testInfoCmp)> seenTests(testInfoCmp);
+        std::set<TestCaseInfo const*, decltype(testInfoCmp) &> seenTests(testInfoCmp);
```
Why is the ref here needed?

objectx replied:
Without `&`, The `std::set` in PlayStation SDK's STL tried to invoke a copy-constructor (and failed).

STL Bug?

```log
MSBuild version 17.4.0+18d5aef85 for .NET Framework
  catch_chronometer.cpp
  catch_benchmark_function.cpp
  catch_run_for_at_least.cpp
  catch_stats.cpp
  catch_get_random_seed.cpp
  catch_tag_alias_autoregistrar.cpp
  catch_test_spec.cpp
  catch_timer.cpp
  catch_tostring.cpp
  catch_totals.cpp
  catch_version.cpp
  catch_generator_exception.cpp
  catch_generators.cpp
  catch_generators_random.cpp
  catch_interfaces_capture.cpp
  catch_interfaces_config.cpp
  catch_interfaces_exception.cpp
  catch_interfaces_registry_hub.cpp
  catch_interfaces_reporter_factory.cpp
  catch_interfaces_reporter_registry.cpp
  catch_interfaces_testcase.cpp
  catch_decomposer.cpp
  catch_errno_guard.cpp
  catch_lazy_expr.cpp
  catch_leak_detector.cpp
  catch_message_info.cpp
  catch_polyfills.cpp
  catch_startup_exception_registry.cpp
  catch_test_case_info_hasher.cpp
  catch_test_case_registry_impl.cpp
  In file included from D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:14:
  In file included from D:\dev\Catch2\src\src\catch2\..\catch2/internal/catch_run_context.hpp:11:
  In file included from D:\dev\Catch2\src\src\catch2\..\catch2/interfaces/catch_interfaces_reporter.hpp:22:
  In file included from C:\tools\SCE\Orbis-SDK\10.000\target\include\map:10:
C:\tools\SCE\Orbis-SDK\10.000\target\include\xtree(1058,11): error : no matching constructor for initialization of 'std::_Tree_comp<false, std::_Tset_traits<const Catch::TestCaseInfo *, (lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28), std::allocator<const Catch::TestCaseInfo *>, false>>::key_compare' (aka '(lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28)') [D:\dev\Catch2\0.B\orbis\src\Catch2.vcxproj]
                  return (key_compare());
                          ^
  C:\tools\SCE\Orbis-SDK\10.000\target\include\xtree(1986,38): note: in instantiation of member function 'std::_Tree_comp<false, std::_Tset_traits<const Catch::TestCaseInfo *, (lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28), std::allocator<const Catch::TestCaseInfo *>, false>>::_Getcomp' requested here
                                  _Addleft = !_DEBUG_LT_PRED(this->_Getcomp(),
                                                                   ^
  C:\tools\SCE\Orbis-SDK\10.000\target\include\xtree(1403,11): note: in instantiation of function template specialization 'std::_Tree<std::_Tset_traits<const Catch::TestCaseInfo *, (lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28), std::allocator<const Catch::TestCaseInfo *>, false>>::_Insert_nohint<const std::_Tree<std::_Tset_traits<const Catch::TestCaseInfo *, (lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28), std::allocator<const Catch::TestCaseInfo *>, false>>::value_type &, std::_Nil>' requested here
                  return (_Insert_nohint(false,
                          ^
  C:\tools\SCE\Orbis-SDK\10.000\target\include\set(226,20): note: in instantiation of member function 'std::_Tree<std::_Tset_traits<const Catch::TestCaseInfo *, (lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28), std::allocator<const Catch::TestCaseInfo *>, false>>::insert' requested here
                  return (_Mybase::insert(_Val));
                                   ^
  D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp(95,41): note: in instantiation of member function 'std::set<const Catch::TestCaseInfo *, (lambda at D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp:88:28)>::insert' requested here
              const auto prev = seenTests.insert( infoPtr );
                                          ^
  D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp(88,28): note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 0 were provided
          auto testInfoCmp = []( TestCaseInfo const* lhs,
                             ^
  D:\dev\Catch2\src\src\catch2\internal\catch_test_case_registry_impl.cpp(88,28): note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 0 were provided
  1 error generated.
```

horenmar replied:
Bleh. This is an upstream bug and needs to be reported there.

I am willing to merge to workaround in the meantime though.

horenmar commented on docs/configuration.md:
```diff
@@ -147,6 +147,7 @@ by using `_NO_` in the macro, e.g. `CATCH_CONFIG_NO_CPP17_UNCAUGHT_EXCEPTIONS`.
     CATCH_CONFIG_WINDOWS_SEH                // Enable SEH handling on Windows
     CATCH_CONFIG_FAST_COMPILE               // Sacrifices some (rather minor) features for compilation speed
     CATCH_CONFIG_POSIX_SIGNALS              // Enable handling POSIX signals
+    CATCH_CONFIG_GETENV                     // System has a working `getenv`
```
I prefer appending to the list, unless some of the options form a logical group.

Also adding new option needs a version placeholder (see `contributing.md` for more).

horenmar commented on docs/configuration.md:
```diff
@@ -162,6 +163,8 @@ Currently Catch enables `CATCH_CONFIG_WINDOWS_SEH` only when compiled with MSVC,
 
 `CATCH_CONFIG_POSIX_SIGNALS` is on by default, except when Catch is compiled under `Cygwin`, where it is disabled by default (but can be force-enabled by defining `CATCH_CONFIG_POSIX_SIGNALS`).
 
+`CATCH_CONFIG_GETENV` is on by default, except when Catch is compiled under some platforms that lacks working `getenv` (currently Windows UWP and Playstation).
```
```suggestion
`CATCH_CONFIG_GETENV` is on by default, except when Catch2 is compiled for
platforms that lacks working `std::getenv` (currently Windows UWP and
Playstation).
```

Catch2 as the project name is preferred in newly (re)written documentation, same for 80 char columns.

horenmar commented on src/catch2/catch_user_config.hpp.in:
```diff
@@ -130,6 +130,16 @@
 
 
+#cmakedefine CATCH_CONFIG_GETENV
+#cmakedefine CATCH_CONFIG_NO_GETENV
```
It also needs to be added to `CMake/CatchConfigOptions.cmake`, so that the option is shown for people using CMake frontends.

horenmar commented on src/catch2/internal/catch_getenv.cpp:
```diff
@@ -14,7 +14,7 @@
     namespace Detail {
 
-#if defined( CATCH_PLATFORM_WINDOWS_UWP )
+#if ! defined (CATCH_CONFIG_GETENV)
```
```suggestion
#if !defined (CATCH_CONFIG_GETENV)
```

just a stylistic nit

horenmar replied:
I can't add a note to unchanged line, so I am adding this one here: you are missing include of `catch_compiler_capabilities` in this file, so the tests break (because the tests no longer can see env variables).

objectx committed:
Adopt suggestions

https://github.com/catchorg/Catch2/pull/2562#discussion_r1016473806

Co-authored-by: Martin Hořeňovský <martin.horenovsky@gmail.com>

objectx committed:
fix: Include catch_compiler_capabilities.hpp explicitly

Suggested in https://github.com/catchorg/Catch2/pull/2562#discussion_r1016478715

objectx committed:
style: Align to project style

Suggested in https://github.com/catchorg/Catch2/pull/2562#discussion_r1016476267

objectx committed:
style: Align to project style guideline

Suggested in https://github.com/catchorg/Catch2/pull/2562#discussion_r1016472204

objectx committed:
fix: Add for CMake frontend

Suggested in https://github.com/catchorg/Catch2/pull/2562#discussion_r1016475142
